### PR TITLE
feat(explorer): active chain as dynamic param

### PR DESCRIPTION
**Pull request by karooolis**
todo:

- [x] Handle layout, error pages, etc.
- [x] Make sure correct chain is configured in production mode
- [x] Throw if non-valid chain is used
- [x] Keep connected between tabs

### CODE REVIEW

File: packages/explorer/src/app/icon.svg
Comment (karooolis) [inline packages/explorer/src/app/icon.svg:1]: Unusual bug but `useParams` would return incorrect params in certain occasions due to favicon not being present i.e. https://github.com/vercel/next.js/discussions/13301 . Not sure if we want to use this as the favicon though.

Comment (holic) [reply packages/explorer/src/app/icon.svg:1]: could also 404 that page with middleware

Comment (karooolis) [reply packages/explorer/src/app/icon.svg:1]: sure but why not just add a favicon? Kind of like the MUD logo as favicon, also shows up in MetaMask wallet when connecting.

Comment (holic) [reply packages/explorer/src/app/icon.svg:1]: we can still add a favicon via meta tag, etc, I just dont love `favicon.ico` specifically haha, it's oooold browser norms and annoying that they still request this by default

and unclear what the expectation is of this file in terms of appropriate format + size

Comment (karooolis) [reply packages/explorer/src/app/icon.svg:1]: I get you, the next.js convention is nice to reduce boilerplate but making it implicit is not so great. Made it more explicit by setting favicon in layout's metadata. Still not ideal but better than before, lmk what you think.

Comment (holic) [reply packages/explorer/src/app/icon.svg:1]: Oh yeah I think we should set a favicon in metadata!

It just doesn't need to be at the exact name/path of `/favicon.ico`. And browsers will attempt to hit that path regardless of what metadata says. That's why I usually just nuke this route in favor of metadata tags with png/svg icons

File: packages/explorer/public/favicon.svg
Comment (holic) [inline packages/explorer/public/favicon.svg:1]: there's another of these svgs somewhere with the paths fixed so they don't have lines between each pixel - can we try to find those?

File: packages/explorer/src/app/[chainName]/(explorer)/worlds/[worldAddress]/page.tsx
Comment (holic) [inline packages/explorer/src/app/[chainName]/(explorer)/worlds/[worldAddress]/page.tsx:1]: can we put chain name param inside the (explorer) dir?

File: packages/explorer/src/common.ts [hunk 1]
```diff
@@ -1,22 +1,36 @@
...
+  [anvil.id]: "anvil",
+  [garnet.id]: "garnet",
+  [redstone.id]: "redstone",
+} as const;
+
+export const chainNameId: Record<SupportedChainNames, SupportedChainIds> = {
+  anvil: anvil.id,
+  garnet: garnet.id,
+  redstone: redstone.id,
+};
```
Comment (holic) [inline packages/explorer/src/common.ts:18]: there's nothing in these types to ensure we maintain parity with the above map

can we compute this or add type constraints to make sure they stay aligned?

File: packages/explorer/src/common.ts [hunk 2]
```diff
@@ -1,22 +1,36 @@
...
+
+export const chains: Record<SupportedChainIds, Chain> = {
   [anvil.id]: anvil,
-  [redstone.id]: redstone,
   [garnet.id]: garnet,
+  [redstone.id]: redstone,
 };
```
Comment (holic) [inline packages/explorer/src/common.ts:24]: alternatively could name the chains here like

```ts
const supportedChains = { redstone, garnet, anvil };
```

then compute the look ups based on that, or do look ups like

```ts
const chain = Object.values(supportedChains).find((c) => c.id === chainId);
```

File: packages/explorer/src/common.ts [hunk 3]
```diff
@@ -1,22 +1,36 @@
...
+    throw new Error(`Invalid chain id. Supported chains are: ${Object.keys(chainIdName).join(", ")}.`);
   }
-
-  return chain;
 }
 
-export function isAnvil() {
-  const chain = getChain();
-  return chain.id === anvil.id;
+export function isValidChainName(name: string | string[] | undefined): asserts name is SupportedChainNames {
```
Comment (holic) [inline packages/explorer/src/common.ts:32]:
```suggestion
export function validateChainName(name: string | string[] | undefined): asserts name is SupportedChainNames {
```

something named `isValid` makes me think this should return a boolean, not throw

File: packages/explorer/src/hooks/useChain.ts [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+export function useChain() {
+  const { chainName } = useParams();
+  isValidChainName(chainName);
+
+  const chainId = chainNameId[chainName];
+  const chain = chains[chainId];
+  return chain;
+}
+
+export function useChainId() {
```
Comment (holic) [inline packages/explorer/src/hooks/useChain.ts:13]: Is this hook necessary?

Comment (karooolis) [reply packages/explorer/src/hooks/useChain.ts:13]: it's a nice helper used in `useContractMutation` and such where we need `chainId` only but not strictly necessary. Do you think it's an overkill?

File: packages/explorer/src/common.ts (hunk 3 above)
Comment (karooolis) [reply packages/explorer/src/common.ts:32]: ah very good point, will do

File: packages/explorer/src/common.ts (hunk 1 above)
Comment (karooolis) [reply packages/explorer/src/common.ts:18]: Initially computed with `reduce` but thought it's too much boilerplate for that. However, your comment made me realize that keeping it aligned is definitely preferred, will change. This is where `_.lodash` is nice but will make without it.

File: packages/explorer/src/hooks/useChain.ts (hunk 4 above)
Comment (holic) [reply packages/explorer/src/hooks/useChain.ts:13]: yeah I find it a bit overkill when you can do 
```ts
const { id: chainId } = useChain();
```

Comment (karooolis) [reply packages/explorer/src/hooks/useChain.ts:13]: makes sense, took inspiration from wagmi but indeed not necessary

File: packages/explorer/public/favicon.svg
Comment (karooolis) [reply packages/explorer/public/favicon.svg:1]: Found one in https://lattice.xyz/brand, looks to have no lines

Comment (holic) [reply packages/explorer/public/favicon.svg:1]: image

🤔 

eh this is fine haha, it'll get shrunk way down

**Review by holic — DISMISSED** [review]

File: packages/explorer/public/favicon.svg
Comment (holic) [reply packages/explorer/public/favicon.svg:1]: oh nm, I see you replaced it with the square one

**Pull request merged**